File: app/src/test/java/com/futsch1/medtimer/schedulerTests/TestHelper.java
package com.futsch1.medtimer.schedulerTests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.futsch1.medtimer.database.FullMedicine;
import com.futsch1.medtimer.database.Medicine;
import com.futsch1.medtimer.database.Reminder;
import com.futsch1.medtimer.database.ReminderEvent;
import com.futsch1.medtimer.reminders.scheduling.ScheduledReminder;

import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;

public class TestHelper {
    private TestHelper() {
        // Hide implicit public constructor
    }

    public static Reminder buildReminder(int medicineId, int reminderId, String amount, int timeInMinutes, int daysBetweenReminders) {
        Reminder reminder = new Reminder(medicineId);
        reminder.reminderId = reminderId;
        reminder.amount = amount;
        reminder.timeInMinutes = timeInMinutes;
        reminder.pauseDays = daysBetweenReminders - 1;
        reminder.consecutiveDays = 1;
        reminder.createdTimestamp = 0;
        reminder.cycleStartDay = 0;
        reminder.days = new ArrayList<>(List.of(true, true, true, true, true, true, true));
        return reminder;
    }

    public static FullMedicine buildFullMedicine(int medicineId, String medicineName) {
        FullMedicine medicineWithReminders = new FullMedicine();
        medicineWithReminders.medicine = new Medicine(medicineName);
        medicineWithReminders.medicine.medicineId = medicineId;
        medicineWithReminders.reminders = new ArrayList<>();
        return medicineWithReminders;
    }

    public static LocalDate on(long day) {
        return LocalDate.ofEpochDay(day - 1);
    }

    public static Instant on(long day, long minutes) {
        return Instant.ofEpochSecond((day - 1) * 86400 + minutes * 60);
    }

    public static Instant onTZ(long day, long minutes, String zoneId) {
        LocalDate localDate = LocalDate.ofEpochDay(day - 1);
        LocalTime localTime = LocalTime.of((int) minutes / 60, (int) minutes % 60);
        return Instant.ofEpochSecond(localDate.atTime(localTime).atZone(ZoneId.of(zoneId)).toEpochSecond());
    }

    public static ReminderEvent buildReminderEvent(int reminderId, long remindedTimestamp) {
        ReminderEvent reminderEvent = new ReminderEvent();
        reminderEvent.reminderId = reminderId;
        reminderEvent.remindedTimestamp = remindedTimestamp;
        return reminderEvent;
    }

    public static void assertReminded(List<ScheduledReminder> scheduledReminders, Instant timestamp, Medicine medicine, Reminder reminder) {
        assertRemindedAtIndex(scheduledReminders, timestamp, medicine, reminder, 0);
    }

    public static void assertRemindedAtIndex(List<ScheduledReminder> scheduledReminders, Instant timestamp, Medicine medicine, Reminder reminder, int index) {
        assertTrue(scheduledReminders.size() > index);
        assertEquals(timestamp, scheduledReminders.get(index).timestamp());
        assertEquals(medicine, scheduledReminders.get(index).medicine().medicine);
        assertEquals(reminder, scheduledReminders.get(index).reminder());
    }

}

File: app/src/test/java/com/futsch1/medtimer/schedulertests/TestHelper.kt
package com.futsch1.medtimer.schedulertests

import com.futsch1.medtimer.model.Medicine
import com.futsch1.medtimer.model.Reminder
import com.futsch1.medtimer.model.ReminderEvent
import com.futsch1.medtimer.model.ReminderTime
import com.futsch1.medtimer.model.ScheduledReminder
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class TestMedicine(medicine: Medicine) {
    private var _medicine: Medicine = medicine
    val reminders: MutableList<Reminder> = mutableListOf()

    val medicine: Medicine get() = _medicine

    var amount: Double
        get() = _medicine.amount
        set(value) {
            _medicine = _medicine.copy(amount = value)
        }

    var expirationDate: LocalDate
        get() = _medicine.expirationDate
        set(value) {
            _medicine = _medicine.copy(expirationDate = value)
        }

    fun toMedicine(): Medicine = _medicine.copy(reminders = reminders.toList())
}

object TestHelper {
    fun buildReminder(
        medicineId: Int,
        reminderId: Int,
        amount: String,
        timeInMinutes: Int,
        daysBetweenReminders: Int
    ): Reminder {
        return Reminder.default().copy(
            id = reminderId,
            medicineRelId = medicineId,
            amount = amount,
            time = ReminderTime(timeInMinutes),
            pauseDays = daysBetweenReminders - 1
        )
    }

    fun buildTestMedicine(medicineId: Int, medicineName: String): TestMedicine {
        return TestMedicine(Medicine.default().copy(id = medicineId, name = medicineName))
    }

    fun on(day: Long): LocalDate {
        return LocalDate.ofEpochDay(day - 1)
    }

    fun on(day: Long, minutes: Long): Instant {
        return Instant.ofEpochSecond((day - 1) * 86400 + minutes * 60)
    }

    fun onTZ(day: Long, minutes: Long, zoneId: String): Instant {
        val localDate = LocalDate.ofEpochDay(day - 1)
        val localTime = LocalTime.of(minutes.toInt() / 60, minutes.toInt() % 60)
        return Instant.ofEpochSecond(
            localDate.atTime(localTime).atZone(ZoneId.of(zoneId)).toEpochSecond()
        )
    }

    fun buildReminderEvent(reminderId: Int, remindedTimestamp: Instant): ReminderEvent {
        return ReminderEvent.default().copy(reminderId = reminderId, remindedTimestamp = remindedTimestamp)
    }

    fun buildReminderEvent(
        reminderId: Int,
        remindedTimestamp: Long,
        reminderEventId: Int
    ): ReminderEvent {
        return ReminderEvent.default()
            .copy(reminderId = reminderId, remindedTimestamp = Instant.ofEpochSecond(remindedTimestamp), reminderEventId = reminderEventId)
    }

    fun assertReminded(
        scheduledReminders: List<ScheduledReminder>,
        timestamp: Instant,
        medicine: Medicine,
        reminder: Reminder
    ) {
        assertRemindedAtIndex(scheduledReminders, timestamp, medicine, reminder, 0)
    }

    fun assertRemindedAtIndex(
        scheduledReminders: List<ScheduledReminder>,
        timestamp: Instant,
        medicine: Medicine,
        reminder: Reminder,
        index: Int
    ) {
        assertTrue(scheduledReminders.size > index)
        assertEquals(timestamp, scheduledReminders[index].timestamp)
        assertEquals(medicine, scheduledReminders[index].medicine)
        assertEquals(reminder, scheduledReminders[index].reminder)
    }
}
